File: IntegrationProject/cpp/src/IntegrationMain.cpp
/*
 * IntegrationMain.cpp
 *
 *  Created on: Dec 28, 2015
 *      Author: chris
 */

#include <iostream>
#include <string>
#include <limits>
#include <sstream>
#include <cctype>
#include "Polynomial.h"
#include "Function.h"
#include "Utilities.h"
#include "Properties.cpp"
#include "IntegrationMethods.h"
#include "IntegrationMethods.cpp"

using namespace std;
using namespace MathFunctions;

// validate user has entered a valid fraction
double validateFraction(string message, string errorMessage) {
   bool isValid = false;
   string fraction;
   double decNum = 0.0;
   do {
      int count = 0;
      cin.clear();
      cout << message << flush;
      cin >> fraction;

      for(int i = 0; i < fraction.size(); i++) {
         if(fraction[i] == '/') {
            count++;
         }
      }

      if(count == 1 || count == 0) {
         isValid = true;
         decNum = Utilities::fractionToDouble(fraction);
      }
      else {
         cin.clear();
         cout << errorMessage << endl;
      }

      if(decNum != decNum || isinf(decNum)) {
         isValid = false;
         cin.clear();
         cout << "\n" + errorMessage + "\n" << endl;
      }

   } while(!isValid);

   return decNum;
}

// validate that user entered a positive integer value
int validatePositiveInt(string message, string errorMessage) {
   bool isValid = false;
   int integerValue = 0;

   do {
      cout << message << flush;
      cin >> integerValue;

      if(cin.good() && integerValue >= 0) {
         isValid = true;
         cin.clear();
         cin.ignore(numeric_limits<streamsize>::max(), '\n');
      }
      else {
         cin.clear();
         cin.ignore(numeric_limits<streamsize>::max(), '\n');
         cout << "\n" + errorMessage + "\n" << endl;
      }
   } while(!isValid);

   return integerValue;
}

// convert polynomial to string
template <typename E, typename C>
string convertPolynomial(Function<E, C> polynomial) {
   stringstream ss;
   ss.clear();
   ss << polynomial;
   return ss.str();
}

int main() {
   int degree = 0;
   int numberOfColumns = 0;
   int optionChosen = 0;
   double coefficient = 0.0;
   double a = 0.0;
   double b = 0.0;
   char continueProgram = 'N';

   Properties props;
   string configFileName = "./config.properties";
   props.read(configFileName);

   cout << props.getValue("WELCOME_TEXT") << endl;

   do {
      // ask user for the degree of the polynomial
      degree = validatePositiveInt("\n" + props.getValue("ASK_FOR_DEGREE"), 
         props.getValue("INVALID_POSITIVE_INTEGER"));

      Polynomial<int, double> newPolynomial(degree);

      cout << endl;

      // ask user for the coefficients of the polynomial for each term
      for(int index = 0; index < newPolynomial.getCoefficients().size(); index++) {
         string stringArray[] = {Utilities::convertInt(index)};
         coefficient = validateFraction(props.getValue("ASK_FOR_COEFFICIENT", 
            stringArray, 1), props.getValue("INVALID_DOUBLE"));
         newPolynomial.setCoefficient(index, coefficient);
      }

      // ask user for starting point of integral
      a = validateFraction("\n" + props.getValue("ASK_FOR_A"), props.getValue("INVALID_DOUBLE"));
      // ask user for the end point of integral
      b = validateFraction("\n" + props.getValue("ASK_FOR_B"), props.getValue("INVALID_DOUBLE"));

      // have user choose the method of integration
      do {
         optionChosen = validatePositiveInt("\n" + props.getValue("CHOOSE_METHOD_MENU"), 
            props.getValue("INVALID_OPTION"));
         if(optionChosen < 1 || optionChosen > 3) {
            cout << props.getValue("INVALID_OPTION") << endl;
         }
      } while(optionChosen < 1 || optionChosen > 3);

      cout << "\n" << endl;

      // depending on user chosen option use that integration method
      double answer;
      switch(optionChosen) {
         case 1: {
            // ask for number of columns
            numberOfColumns = validatePositiveInt("\n" + props.getValue("ASK_FOR_N"), 
               props.getValue("INVALID_POSITIVE_INTEGER"));
            answer = IntegrationMethods::rectangleMethod(a, b, numberOfColumns, newPolynomial);
            const string stringArray[] = {convertPolynomial(newPolynomial), 
               Utilities::fraction(a, 0.0001), Utilities::fraction(b, 0.0001), 
               Utilities::convertDouble(answer), Utilities::convertInt(numberOfColumns)};
            cout << "\n" + props.getValue("SOLVE_RECTANGLE", stringArray, 5) << endl;
            break;
         }
         case 2: {
            // ask for number of columns
            numberOfColumns = validatePositiveInt("\n" + props.getValue("ASK_FOR_N"), 
               props.getValue("INVALID_POSITIVE_INTEGER"));
            answer = IntegrationMethods::trapezoidalMethod(a, b, numberOfColumns, newPolynomial);
            const string stringArray[] = {convertPolynomial(newPolynomial), 
               Utilities::fraction(a, 0.0001), Utilities::fraction(b, 0.0001), 
               Utilities::convertDouble(answer), Utilities::convertInt(numberOfColumns)};
            cout << "\n" + props.getValue("SOLVE_TRAPEZOIDAL", stringArray, 5) << endl;
            break;
         }
         case 3: {
            answer = IntegrationMethods::symbolicMethodForPolynomial(a, b, newPolynomial);
            const string stringArray[] = {convertPolynomial(newPolynomial), 
               Utilities::fraction(a, 0.0001), Utilities::fraction(b, 0.0001), 
               Utilities::convertDouble(answer)};
            const string stringArrayIntegral[] = {IntegrationMethods::
               integralToString(newPolynomial)};
            cout << "\n" + props.getValue("SOLVE_SYMBOLIC", stringArray, 4) << endl;
            cout << "\n" + props.getValue("SHOW_INTEGRAL", stringArrayIntegral, 1) << endl;
            break;
         }
         default:
            cout << "\n" + props.getValue("INVALID_OPTION") << endl;
      }

      // ask user if they want to integrate another polynomial
      do {
         cout << "\n" + props.getValue("CONTINUE_PROGRAM") << endl;
         cin >> continueProgram;
      } while(continueProgram != 'Y' && continueProgram != 'N'
           && continueProgram != 'y' && continueProgram != 'n');

   } while(continueProgram == 'Y' || continueProgram == 'y');

   return 0;
}

File: IntegrationProject/java/src/com/project/math/IntegrationMain.java
package com.project.math;
import java.text.MessageFormat;
import java.text.ParseException;
import java.util.InputMismatchException;
import java.util.Scanner;

import com.project.math.utilities.IntegrationMethods;
import com.project.math.utilities.PropertiesFile;
import com.project.math.utilities.Utilities;

/**
 * This is the driver for the project this includes main
 */
public class IntegrationMain {
   public static void main(String[] args) {
      Scanner keyboard = new Scanner(System.in);
      PropertiesFile propertiesFile = new PropertiesFile();

      int degree = -1;
      double a = 0;
      double b = 0;
      int n = 0;
      String continueProgram = "";
      
      // display welcome text
      System.out.println(propertiesFile.getPropertyValue("WELCOME_TEXT"));
      do {
         Function<Integer, Double> polynomial = new Polynomial<Integer, Double>();

         // receive the degree of the polynomial from user
         do {
            degree = Utilities.validateInt(keyboard, "\n" 
               + propertiesFile.getPropertyValue("ASK_FOR_DEGREE"), 
            propertiesFile.getPropertyValue("INVALID_POSITIVE_INTEGER"));
         } while(degree < 0);

         // create the exponents for the polynomial
         for(int index = degree; index >= 0; index--) {
            polynomial.addToExponent(index);
         }

         // ask for all coefficients from the user
         int index = 0;
         MessageFormat messageFormat = new MessageFormat(propertiesFile
            .getPropertyValue("ASK_FOR_COEFFICIENT"));
         while(index <= degree) {
            double coe;
            String[] stringVariable = {"" + index};
            coe = Utilities.validateDouble(keyboard, messageFormat.format((stringVariable)), 
               propertiesFile.getPropertyValue("INVALID_DOUBLE"));
            polynomial.addToCoefficient(0, coe);
            index++;
         }

         // ask for the starting point of integral (a)
         a = Utilities.validateDouble(keyboard, "\n" + propertiesFile.getPropertyValue("ASK_FOR_A"), 
            propertiesFile.getPropertyValue("INVALID_DOUBLE"));
         // ask for the ending point of integral (b)
         b = Utilities.validateDouble(keyboard, "\n" + propertiesFile.getPropertyValue("ASK_FOR_B"), 
            propertiesFile.getPropertyValue("INVALID_DOUBLE"));

         // choose the method to integrate the solution
         int choosenMethod = 0;
         messageFormat = new MessageFormat("\n" + propertiesFile
            .getPropertyValue("ASK_FOR_INTEGRATION_METHOD"));
         String[] stringArray = {polynomial.toString()};
         do {
            choosenMethod = Utilities.validateInt(keyboard, messageFormat.format(stringArray) 
               + "\n" + propertiesFile.getPropertyValue("CHOOSE_METHOD_MENU"), 
               propertiesFile.getPropertyValue("INVALID_POSITIVE_INTEGER"));
         } while(choosenMethod <= 0 || choosenMethod >= 4);

         // depending on what user chooses calculate the integral and display output
         // the string array converts the parameters to strings in order to add
         // the information to the message to display to user
         double answer = 0.0;
         String[] stringArrayParams;
         switch(choosenMethod) {
            case 1:
               do {
                  // aks user for the number of columns (n)
                  n = Utilities.validateInt(keyboard, "\n" + propertiesFile
                     .getPropertyValue("ASK_FOR_N"), propertiesFile.getPropertyValue(
                     "INVALID_POSITIVE_INTEGER"));
               } while(n <= 0);
               answer = IntegrationMethods.rectangleMethod(a, b, n, polynomial);
               stringArrayParams = new String[]{polynomial.toString(), "" + a, "" + b, "" 
                  + answer, "" + n};
               messageFormat = new MessageFormat(propertiesFile.getPropertyValue("SOLVE_RECTANGLE"));
               System.out.println("\n" + messageFormat.format(stringArrayParams));
               break;
            case 2:
               do {
                  // ask user for the number of columns (n)
                  n = Utilities.validateInt(keyboard, "\n" + propertiesFile
                     .getPropertyValue("ASK_FOR_N"), propertiesFile.getPropertyValue(
                     "INVALID_POSITIVE_INTEGER"));
               } while(n <= 0);
               answer = IntegrationMethods.trapezoidalMethod(a, b, n, polynomial);
               stringArrayParams = new String[]{polynomial.toString(), "" + a, "" + b, "" 
               + answer, "" + n};
               messageFormat = new MessageFormat(propertiesFile.getPropertyValue(
                  "SOLVE_TRAPEZOIDAL"));
               System.out.println("\n" + messageFormat.format(stringArrayParams));
               break;
            case 3:
               answer = IntegrationMethods.symbolicMethod(a, b, polynomial);
               stringArrayParams = new String[]{polynomial.toString(), "" + a, "" + b, "" + answer};
               messageFormat = new MessageFormat("\n" + propertiesFile
                  .getPropertyValue("SOLVE_SYMBOLIC"));
               System.out.println(messageFormat.format(stringArrayParams));
               stringArrayParams = new String[]{IntegrationMethods.integralToString(polynomial)};
               messageFormat = new MessageFormat(propertiesFile.getPropertyValue(
                  "SHOW_INTEGRAL"));
               System.out.println(messageFormat.format(stringArrayParams));
               break;
            default:
               System.out.println("\n" + propertiesFile.getPropertyValue("INVALID_POSITIVE_INTEGER"));
               break;
            }

            do {
               System.out.println("\n" + propertiesFile.getPropertyValue("CONTINUE_PROGRAM"));
               continueProgram = keyboard.nextLine();
         } while(!continueProgram.equalsIgnoreCase("Y") && !continueProgram.equalsIgnoreCase("N"));
      } while(continueProgram.equalsIgnoreCase("Y"));
   keyboard.close();
   }
}
